### DISCUSSION TITLE
to_set() function

### DISCUSSION THREAD

**Discussion by innovate-invent** [jmespath-community/jmespath.spec#45] [Ideas]
We need a mechanism to reduce an array to its unique elements.

The proposed function signature would be `array to_set(array arg)`.

The first occurrence of each element will be kept in the order it is encountered. Elements are compared with each types equivalence logic.

example:
`search([1,1,2,1,4], to_set(@))` -> `[1,2,4]`
`search(["a", {"a":1}, {"a":1, "b":2}, [1],{"a":1}, 3, [2]], to_set(@))` -> `[["a", {"a":1}, {"a":1, "b":2}, [1], 3, [2]]`

I chose "to_set" to be consistent with the other "to_*" functions, but this could alternately be named "unique()" or "deduplicate()".

**Comment by springcomp** [comment]
That makes me think that we may want something like `fold()` in the functional programming world...

**Reply by innovate-invent** [reply]
I was thinking of that myself, but the `merge()`, flatten `[]` and `map()` operations can be used to cover most of its cases.

**Reply by springcomp** [reply]
Oh, interesting. That needs a new entry in the [Frequently Used Paths](https://github.com/jmespath-community/jmespath.site/wiki/Frequently-used-paths) then 😏.

Maybe this should be moved here or promoted to full blown site page...

**Reply by innovate-invent** [reply]
I am actually having a hard time imagining a use case for fold. Did you have any in mind?

**Reply by springcomp** [reply]
Together with simple arithmetic operators, I was thinking about the usual summation of values. For instance, to calculate the total cost of cloud resources out of a JSON report of billing usage.

Not sure if it's a real world scenario, that's more a thought experiment at this stage.

**Reply by innovate-invent** [reply]
Ah, I see, I wasn't thinking of folding primitives. In that case you are right, let me create another discussion for this.

**Comment by springcomp** [comment]
I think functions prefixed with `to_` [should be reserved](https://github.com/jmespath/jmespath.py/pull/105#issuecomment-199362038) to those that accept `any` type.

Your suggestion makes me think of a larger use case. Maybe this particular use case could be addressed with a `sort()` function.

```
array sort(array $subject [, bool $unique])
```

```
search([1,1,2,1,4], sort(@)) -> [1,1,1,2,4]
search([1,1,2,1,4], sort(@, `true`)) -> [1,2,4]
```

To address other use cases, such as sorting in reverse order, maybe the `$unique` argument should be a string accepting a combination of specifiers with values such as `u` and `r`, etc.

**Reply by springcomp** [reply]
Incidentally, if we agree to the naming conventions, we may want to specify this explicitely in the spec.

**Reply by innovate-invent** [reply]
I see what you mean about the `to_` prefix. I am not sure if I like the idea that the items would have to be sorted to get a set of unique items, I might want to keep the order in some cases.

Adding a `unique` arg to the sort function doesn't seem fitting for the purpose of sort either.

The `reverse()` function should also be used rather than passing an argument to sort.

**Reply by innovate-invent** [reply]
I think we should keep the `to_` prefix and expand the definition of this function to accept any. This would result in an implicit to_array() call if an array is not passed as the argument.

**Reply by springcomp** [reply]
So it would behave on a scalar value as if it were invoked with an array containing a single argument ?

**Reply by innovate-invent** [reply]
Exactly, it would behave exactly as [to_array](https://jmespath.org/specification.html#to-array), but de-duplicate values.

**Reply by springcomp** [reply]
I’m not a strong proponent of side-effects like these. I would rather enforce the type constraint like `avg()` or `map()` do for instance.

**Comment by hell-racer** [comment]
I'd just call this `distinct()` :)
(I like `unique()` too)

**Reply by springcomp** [reply]
Python calls this `set()`. Should we use that instead ?

- `` search( set(@), [1, 2, 4, 3, 2, 1] ) -> [1, 2, 4, 3] ``

**Reply by jdevillard** [reply]
For me also , distinct or unique is more explicit than set !

**Reply by innovate-invent** [reply]
I think I am going to vote for 'distinct' if not 'to_set'

**Reply by springcomp** [reply]
Let's settle on `distinct()` then. 👍
